BDI Enhancements

davidhabib opened the pull request:
# Critical Changes

## Batch Data Import Enhancements

To support the new Batch Data Import feature, you need to perform some setup tasks. In the profiles of all users who perform data imports using the NPSP Data Importer:

- Set the NPSP Data Import Batch tab setting to Default On.
- Set object permission for the NPSP Data Import Batch object to Read, Create, Edit.
- In the NPSP Data Import object settings, set field permissions for the NPSP Data Import Batch field to Read Access and Edit Access.
- Override the Clone, Edit, and New buttons on the NPSP Data Import Batch object to use the BDI_BatchOverride Visualforce page.

 
# Changes

## Batch Data Import Enhancements

- The NPSP Data Importer now includes support for processing your NPSP data imports in batches, using a specific set of configuration settings. Use the new NPSP Data Import Batch object to create a batch and specify its unique configuration settings.
- You can also now schedule NPSP Data Import Batches to be processed automatically using a scheduled job. You can find the new scheduled job, **NPSP 09 - Data Import Batch Processing**, in Setup under Scheduled Jobs. NPSP will automatically schedule the new NPSP 09 - Data Import Batch Processing job, but the default "Nonprofit Success Pack" user has limited permissions. To ensure smooth functioning of scheduled imports, the job should be scheduled by a System Administrator. A System Administrator can automatically reschedule the job by simply visiting the NPSP Settings page.
- The NPSP Data Importer now supports invoking a custom Post Process Apex class to perform additional tasks as each set of Data Import records is processed. This feature requires working with a developer to write an Apex class that implements the new BDI_IPostProcess interface and its method. You then specify the Apex class in the **Post Process Implementing Class** field in either the NPSP Data Import configuration options or in the NPSP Data Import Batch configuration options.
- For complete documentation, see the NPSP Advanced User's Guide to Importing Data (https://powerofus.force.com/articles/Resource/NPSP-Advanced-User-s-Guide-to-Importing-Data).

# Issues Closed

davidhabib commented:
@smarovich Can you please review and update this pull request comment to capture all of the new features in BDI that we want to hilight, as well as pointing people at your new advanced guide.  @judisohn did i forget any of the new features worth highlighting?  thanks both of you!

judisohn commented:
@smarovich @davidhabib This covers the main features! We may want to include similar language for the post processing as we did for the custom household naming to guide developers in the right direction in release notes. Should we also have language advising the enabling of the new object/new fields for push upgrades?

davidhabib commented:
@judisohn @smarovich I've updated the critical changes comments to include the items that need enabling.  please cleanup to use appropriate language as you see fit.  @njjc I've added a new night batch job.  Do you recall if our upgrade script will add new jobs, or whether we need to tell people to open the NPSP Settings page to cause the jobs to get recreated?

njjc commented:
We have a bit of a two-pronged approach, we do schedule it in the post install script, but it's scheduled by the `Nonprofit Success Pack` user which as limited permissions and can hit some errors. Then, the first time a SysAdmin accesses the Settings page, it will reschedule the job with that user. I don't think we need to mention it in the release notes.

judisohn commented:
@davidhabib @njjc The doc @smarovich talks about looking for and verifying that scheduled job. I think we can leave it at that for now. On the Hub post, we should be sure to link to the new doc.

njjc commented on src/classes/BDI_SettingsUI_CTRL.cls:
```diff
@@ -133,45 +125,36 @@ public with sharing class BDI_SettingsUI_CTRL {
+    * @return list<SelectOption>
     */
-    public static boolean saveBDISettings(Data_Import_Settings__c diSettings) {
+    private list<SelectOption> listSOForObject(String strObject) {
```
I take it you're not using dfr.isExternalID() in case they're using a field that's not marked as external ID?

I'm also wondering if it's possible that a unique ID field could be packaged at some point, are you sure showing only unpackaged fields makes sense?

davidhabib replied:
yes, the field does not have to be an official external ID.  As for packaged fields from another package, I entered issue #2700 several weeks ago that BDI does not correctly support them.

njjc commented on src/classes/UTIL_MasterSchedulableHelper.cls:
```diff
@@ -71,7 +71,8 @@ public without sharing class UTIL_MasterSchedulableHelper {
         'NPSP 07 - Seasonal Address Updates'        => new List<String>{'0 0 23 ? * *', 'ADDR_Seasonal_SCHED'},
-        'NPSP 08 - Level Assignment Updates'        => new List<String>{'0 0 4 ? * *',  'LVL_LevelAssign_SCHED'}
+        'NPSP 08 - Level Assignment Updates'        => new List<String>{'0 0 4 ? * *',  'LVL_LevelAssign_SCHED'},
+        'NPSP 09 - Data Import Batch Processing'    => new List<String>{'0 0 22 ? * *', 'BDI_DataImportBatch_SCHED'}
```
I think it might be worth testing what happens when trying to run this with the `Nonprofit Success Pack` packaging user with limited access. If it borks, which seems likely, it makes sense to update the docs to suggest visiting the NPSP settings page to get it rescheduled. What do you think?

davidhabib replied:
yes, I believe sarah's doc does tell people to verify the job, and if it isn't there, to go to the settings page.

njjc commented on src/classes/BDI_BatchOverride_CTRL.cls:
```diff
@@ -0,0 +1,104 @@
+        try {
+            BDI_DataImportService.validateSettings(diBatch);
+            upsert diBatch;
+            return new PageReference('/' + diBatch.Id);
```
This is definitely broken in LEX. When clicking Save, I get a blank page with this popup:
![screen shot 2017-09-07 at 10 34 31 am](https://user-images.githubusercontent.com/1018198/30176738-60e6a88a-93b8-11e7-833c-8a44e6fe20d7.png)
Both options are broken.
- If I click Cancel, it sends me back to edit the UI with everything populated, but the Batch object has been created. 
- If I click Open, it opens a new tab on the newly created Batch, but the old tab still has the Edit UI with everything I filled out populated.
In both cases, I can then click Save again and it creates a duplicate Batch.

jjbennett replied:
I would use the javascript hybrid approach, we have the UTIL_NavigateBack component that will refresh the page, not sure if that will help with the popup. If you enable popups for your org than you don't get that.

njjc commented on src/layouts/DataImportBatch__c-NPSP Data Import Batch Layout.layout:
```diff
@@ -1,5 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
The default button options on the page layout are weird. Any idea if we can update the defaults to show Edit, Clone, and Process Batch instead of all those weird New (Object) buttons?
![screen shot 2017-09-07 at 10 38 06 am](https://user-images.githubusercontent.com/1018198/30176934-fe38f16a-93b8-11e7-8fd0-58b675c78e46.png)

davidhabib replied:
I've just updated the quick action section to put Edit, Clone, Process Batch, and Delete first.  I left the others alone since they are Salesforce defaults, and I really didn't know what was best.  good catch by the way!

njjc commented on src/objects/DataImportBatch__c.object:
```diff
@@ -173,28 +190,123 @@
+        <masterLabel>Process Batch</masterLabel>
+        <openType>sidebar</openType>
+        <protected>false</protected>
+        <url>/apex/npsp__BDI_DataImport?batchId={!DataImportBatch__c.Id}&amp;retURL={!URLFOR(&apos;/&apos; + DataImportBatch__c.Id)}</url>
```
I'm unable to test this Process Batch button in LEX due to the namespace, we should have this tested in a beta package or during regression.

njjc commented on src/classes/BDI_DataImportService.cls:
```diff
@@ -0,0 +1,2252 @@
+
+        // we need the Contact CustomId key to always be checked second when matching
+        string strUniqueId = '';
+        if (isCustomIDInContactMatchRules && (strUniqueId = strNull(con.get(diSettings.Contact_Custom_Unique_ID__c))) != '') {
```
This is confusing to me. It looks like strUniqueId is getting a Boolean value.

njjc replied:
I see you're doing the same thing in a bunch of places. I guess it's fine ;)

davidhabib replied:
sorry it isn't clear.  it's assigning the unique Id to the variable (so it can be used again later), and then comparing to ''.

jjbennett commented on src/pages/BDI_BatchOverride.page:
```diff
@@ -0,0 +1,56 @@
+                            <div class="slds-media__figure">
+                              <span class="slds-icon_container slds-icon-custom-custom18">
+                                <svg class="slds-icon" aria-hidden="true">
+                                  <use xlink:href="{!URLFOR($Asset.SLDS, 'assets/icons/custom-sprite/svg/symbols.svg#custom17')}" ></use>
```
The SVGs have an issue in Edge when in Lightning and in IE11, I switched to using the .pngs instead. I also created a header component to just pass the png name and button actions for reusability but it's not merged yet, jess is currently testing it.

davidhabib commented:
@njjc I think this pull request is ready again.  Thanks to @jjbennett for doing the LEX nav fixes.  I've updated the page layout.  please let me know if there is anything else that needs to be done.

As for the globals, I've reviewed them again, and I'm good with moving forward with them.  Basically, I only made global the most minimal set of properties and methods that our implementation of IMatchDonation needed to use.  This is a pretty small set, and I believe they do make sense to expose to customizers, and I don't believe they will prevent us from further evolution of the BDI_DataImportService in the future.

njjc commented:
Everything looks great, I agree that the globals are looking good.

My only concern, as mentioned before, is that the scheduled imports won't work with the "Nonprofit Success Pack" user that will schedule the job during the push upgrade's post install script. I think @smarovich should update the release notes to say something along the lines of "Verify that the job is scheduled by a System Administrator, not the `Nonprofit Success Pack` user."

davidhabib commented:
@judisohn @smarovich please see Nic's comment about the scheduling of the job.  We may need a clarification in the doc.  thanks.

smarovich commented:
@davidhabib Nic and I discussed his comments and added to the release notes. I'll add it to the docs today too. Thanks!

njjc commented on src/classes/BDI_DataImportService.cls:
```diff
@@ -0,0 +1,2252 @@
+    /*******************************************************************************************************
+    * @description constructor
+    */
+    public BDI_DataImportService() {
```
I'm wondering if this constructor needs to be global so it can be constructed by an external caller.

njjc replied:
The instance of this class is passed fully constructed to the caller.

njjc commented on src/classes/BDI_DataImportService.cls:
```diff
@@ -0,0 +1,2252 @@
+    /*******************************************************************************************************
+    * @description holds the list of DataImport records we are currently processing
+    */
+    global list<DataImport__c> listDI;
```
Should these global vars be moved further up so they're more immediately apparent?

njjc commented on src/classes/BDI_IMatchDonations.cls:
```diff
@@ -47,6 +47,6 @@ public interface BDI_IMatchDonations {
     * @return void
     */ 
-    void match(BDI_DataImport_BATCH bdi, list<DataImport__c> listDI, 
+    void match(BDI_DataImportService bdi, list<DataImport__c> listDI,
```
I think we should add global here for clarity, even if it's implicit.

njjc replied:
Adding global isn't possible with an interface, apparently.

njjc commented on src/classes/BDI_IPostProcess.cls:
```diff
@@ -0,0 +1,45 @@
+    * @param bdi The BDI instance currently running
+    * @return void
+    */
+    void process(BDI_DataImportService bdi);
```
I think we should add global here for clarity, even if it's implicit.

njjc replied:
Adding global isn't possible with an interface, apparently.
